### PR TITLE
Create a Food Recipe App

### PR DESCRIPTION
**Pull request by zunairkhan811**
### The Project is completed according to Project Requirements:

**The Project has two Team Members:**
1- Muhammad Zunair khan(@zunairkhan811 )
2- Lawmsangi Varte (@Lawmsangi )

 **Project requirements:**
  - You should follow the layout of the wireframes provided. You should personalize the rest of the design including colors, typographies, spacings, etc.

  - Login page and registration page:

  - Should be built with Devise.
 
- Recipes list:

  - Should display a list of recipes created by the logged-in user as in the wireframe.
  - Should lead to recipe details.
  
- Public recipe list:

  - Should display a list of all public recipes ordered by newest as in the wireframe.
  - Should lead to recipe details.
  - If the user is the owner of the recipe, should allow the user to delete it.
  
- Recipe details:

  - Should display a toggle button that allows for a recipe to be made public or private.
  - If the recipe is public or the user is the owner of the recipe, should display the recipe details as in the wireframe.
  - If the user is the owner of the recipe, should lead to the form that allows the user to add new food.
  
- Make sure there are no N+1 queries happening.

- Create a navigation menu that allows users to open all of the pages you created.

- Write unit and integration tests

**Only for a group that has less than 3 members:**

- Food list:
  - Should display a list of food added by the logged-in user as in the wireframe (display also quantity of a given food).
  - Should lead to a form that allows users to add new food.
- General shopping list view:
  - Should show the list of food that is missing for all recipes of the logged-in user (compare the list of food for all 
     recipes with the general food list of that user).
  - Should count the total food items and total price of the missing food.

### CODE REVIEW

File: app/controllers/recipes_controller.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      @recipe.update(public: 1)
+      redirect_to @recipe, notice: 'Recipe public status was successfully updated.'
+    elsif @recipe.public == 1
+      @recipe.update(public: 0)
+      redirect_to @recipe, notice: 'Recipe public status was successfully updated.'
+    else
+      render :edit
+    end
+  end
+
```
Comment (Munsa1) [inline app/controllers/recipes_controller.rb]: - [x] Any form of manipulation of a recpie should be done by the user who created said recipe. Right now a user that did not create a recipe is able to change it from public to private, causing a security problem kindly fix this.

Video cast
[cinnamon-2023-12-22T171056+0200.webm](https://github.com/zunairkhan811/food-recipe/assets/81932478/1ecbdda0-da07-4188-bc1d-40ae5e92b926)

File: app/controllers/recipes_controller.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  def destroy
+    @recipe = Recipe.find(params[:id])
+    authorize! :destroy, @recipe
+    if @recipe.destroy
+      redirect_to recipes_path, notice: 'Recipe deleted'
+    else
+      redirect_to recipes_path, alert: 'Recipe could not be deleted'
+    end
+  end
```
Comment (Munsa1) [inline app/controllers/recipes_controller.rb:48]: - [ ] We should that only the user that created a food is able to delete said food for improved security of the app

Videocast
[cinnamon-2023-12-22T171604+0200.webm](https://github.com/zunairkhan811/food-recipe/assets/81932478/26905854-47c2-404d-8036-8e4925d5f4d9)

File: app/views/recipes/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    </div>
+  </div>
+
+  <table class="table mt-4">
+    <thead>
+      <tr>
+        <th>Food</th>
+        <th>Quantity</th>
+        <th>Value</th>
+        <th>Actions</th>
```
Comment (Munsa1) [inline app/views/recipes/show.html.erb:44]: - [x] On The recipe show page we should be able to generate a shopping list. Kindly fix this issue.


Videocast
[cinnamon-2023-12-22T172824+0200.webm](https://github.com/zunairkhan811/food-recipe/assets/81932478/4da18548-899a-4f41-bfe5-c135e5641577)

**Review by Munsa1 — CHANGES_REQUESTED** [review]
Hi @zunairkhan811 @Lawmsangi, 

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>


## Suggested changes
_Check the comments under the review._

You can use as many of my suggestions as you want. If there is anything you would like to skip - feel free to do that. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

File: app/controllers/recipes_controller.rb (hunk 1 above)
Comment (zunairkhan811) [reply app/controllers/recipes_controller.rb]: @Munsa1  Thanks for notifying me. The changes have been made.

File: app/controllers/recipes_controller.rb (hunk 2 above)
Comment (zunairkhan811) [reply app/controllers/recipes_controller.rb:48]: @Munsa1 Thanks for your suggestions. I have made the changes as said by you.

File: app/views/recipes/show.html.erb (hunk 3 above)
Comment (zunairkhan811) [reply app/views/recipes/show.html.erb:44]: @Munsa1 Thanks for your suggestions. So that you know – the changes have been made as you said.

**Review by OyePriscilla — APPROVED** [review]
Hi @zunairkhan811 and @Lawmsangi 👏 👏 👏 👏, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights
Implemented authorization for the app ✔️ 
Generating shopping list implemented ✔️ 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**